lnode/synctxnpool.go

```diff
@@ -0,0 +1,349 @@
+package lnode
+
+import (
+	"bytes"
+	"encoding/hex"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/nknorg/nkn/v2/chain/pool"
+	"github.com/nknorg/nkn/v2/common"
+	"github.com/nknorg/nkn/v2/config"
+	"github.com/nknorg/nkn/v2/node"
+	"github.com/nknorg/nkn/v2/pb"
+	"github.com/nknorg/nkn/v2/transaction"
+	"github.com/nknorg/nkn/v2/util"
+	"github.com/nknorg/nkn/v2/util/log"
+)
+
+// Number of random neighbors to sync txn pool
+const NumNbrToSyncTxnPool = 1
+
+func (localNode *LocalNode) StartSyncTxnPool() {
+	for {
+		dur := util.RandDuration(config.ConsensusDuration/2, 0.4)
+		time.Sleep(dur)
+		localNode.SyncTxnPool()
+	}
+}
+
+// sync txn pool process
+func (localNode *LocalNode) SyncTxnPool() {
+
+	// remove expired twt before sending request
+	localNode.TxnPool.RemoveExpiredTwt()
+
+	neighbors := localNode.GetNeighbors(nil)
+	neighborCount := len(neighbors)
+
+	if neighborCount <= 0 { // don't have any neighbor yet
+		return
+	}
+
+	if NumNbrToSyncTxnPool < neighborCount { // choose random neighbors
+
+		r := rand.New(rand.NewSource(time.Now().UnixNano()))
+		r.Shuffle(len(neighbors), func(i, j int) { neighbors[i], neighbors[j] = neighbors[j], neighbors[i] })
+
+		neighbors = neighbors[:NumNbrToSyncTxnPool]
+	}
+
+	myHash, _ := localNode.TxnPool.GetTwtFingerprintAndCount()
+	maxCount := uint32(0)
+	var neighborToSync *node.RemoteNode // the neighbor which is sync txn target
+	var wg sync.WaitGroup
+
+	for _, neighbor := range neighbors {
+		wg.Add(1)
+
+		go func(nbr *node.RemoteNode) {
+			defer wg.Done()
+
+			replyMsg, err := localNode.requestTxnPoolHash(nbr)
+			if err != nil {
+				log.Warningf("sendReqTxnPoolHash neighbor %v error: %v", nbr.GetAddr(), err)
+				return
+			}
+			log.Infof("Get reply TxnPoolHashAndCount, neighbor's txn count is: %v, hash is: %v, my hash: %v ",
+				replyMsg.TxnCount, hex.EncodeToString(replyMsg.PoolHash)[:8], hex.EncodeToString(myHash)[:8])
+
+			if !bytes.Equal(replyMsg.PoolHash, myHash) {
+				if replyMsg.TxnCount > maxCount { // Choose the neighbor which has most txn count in txn pool
```

Comment:
This if need to be protected by a lock to prevent race condition caused by multiple goroutines

Reply:
Yes, a lock is added.